### PR TITLE
feat/EN-7034 (1/2)

### PR DESCRIPTION
**Pull request by Jvillegasd**
Changelog:

- Data access layer for snowflake database connection
- Models created
- Schemas created
- Process lasted pre approval outputs algorithm
- Store batches in dynamodb

TODO:
- In next pr we'll include the schedule part of the requeriment

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+pytest
```
Comment (CFCarlo1911) [inline requirements-dev.txt:1]: I think you can add this line and just run requirements-dev.txt for both
```suggestion
-r requirements.txt
pytest
```

File: conftest.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+import os
```
Comment (CFCarlo1911) [inline conftest.py:1]: I think you can use isort:
```suggestion
import os
from typing import (
    Iterator,
    TYPE_CHECKING
)

import boto3
import pytest
from moto import mock_dynamodb
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from src.connection.dal import DataAccessLayer, Base
```

File: src/connection/dal.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import os
```
Comment (CFCarlo1911) [inline src/connection/dal.py:1]:
```suggestion
import os
from contextlib import contextmanager
from typing import Iterator

from snowflake.sqlalchemy import URL
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm.session import Session

```

File: src/handler.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    SalesforceBatchRegistry,
+    LastSalesforceBatchRegistry
+)
+from src.schemas.workflow import WorkflowOutput
+from src.models.pre_approval_output import PreApprovalOutput
+
+dal: DataAccessLayer = DataAccessLayer()
+dynamodb_res = boto3.resource('dynamodb')
+batch_table = dynamodb_res.Table('SnowflakeBatchRegistryTable')
+last_batch_table = dynamodb_res.Table('LastSalesforceBatchTable')
```
Comment (CFCarlo1911) [inline src/handler.py:17]: I think the table name could be an environment variable 🤔

File: src/handler.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    if last_sf_batch:
+        current_sf_sync_batch += last_sf_batch.sync_batch
+
+    current_sf_batch = SalesforceBatchRegistry(
+        dynamodb_table=batch_table,
+        loanapp_ids=[],
+        failed_loanapp_ids=[],
+        sync_batch=current_sf_sync_batch
+    )
+    with dal.get_session() as db:
```
Comment (CFCarlo1911) [inline src/handler.py:53]: Why `dal` is a global variable?

File: src/handler.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        LastSalesforceBatchRegistry.save_batch(
+            dynamodb_table=last_batch_table,
+            next_sf_batch=processed_batch
+        )
+        LastSalesforceBatchRegistry.remove_batch(
+            dynamodb_table=last_batch_table,
+            batch_id=processed_output['last_sf_batch_id']
+        )
+    else:
+        response_msg = 'There are not exists new pre approval outputs...'
```
Comment (CFCarlo1911) [inline src/handler.py:80]:
```suggestion
        response_msg = 'There are no existing new pre-approval outputs...'
```

File: src/models/batch_registry.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
+import uuid
```
Comment (CFCarlo1911) [inline src/models/batch_registry.py:1]:
```suggestion
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, TYPE_CHECKING
```

File: src/models/pre_approval_output.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import sys
```
Comment (CFCarlo1911) [inline src/models/pre_approval_output.py:1]:
```suggestion
import sys

from snowflake.sqlalchemy import ARRAY
from sqlalchemy import (
    Column,
    String,
    Integer,
    Float,
    TIMESTAMP
)

from src.connection.dal import Base
```

File: src/models/pre_approval_output.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    TIMESTAMP
+)
+from snowflake.sqlalchemy import ARRAY
+from src.connection.dal import Base
+
+
+class PreApprovalOutput(Base):
+    __tablename__ = 'PRE_APPROVAL_OUTPUT_TEST'
+
+    LOANAPP_ID = Column(Integer)
```
Comment (CFCarlo1911) [inline src/models/pre_approval_output.py:16]: Why the columns are in UPPER_CASE?

File: tests/factories/pre_approval.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import pytz
+import factory
+import factory.fuzzy
+from datetime import datetime, timedelta, timezone
+from src.models.pre_approval_output import PreApprovalOutput
+
+NUMBER_OF_BALANCES: int = 5
+now = datetime.now(tz=timezone.utc)
```
Comment (CFCarlo1911) [inline tests/factories/pre_approval.py:8]: If it's a constant I recommend to use UPPER_CASE:
`NOW = datetime.now(tz=timezone.utc)`

File: tests/factories/pre_approval.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+next_one_week = now + timedelta(days=7)
```
Comment (CFCarlo1911) [inline tests/factories/pre_approval.py:9]: Same :)

File: tests/test_handler.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
+import json
```
Comment (CFCarlo1911) [inline tests/test_handler.py:1]: import json
from typing import TYPE_CHECKING

from src.connection.dal import DataAccessLayer
from src.schemas.workflow import WorkflowOutput
from tests.factories.pre_approval import PreApprovalFactory

File: tests/test_local_db.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
+from typing import List, TYPE_CHECKING
```
Comment (CFCarlo1911) [inline tests/test_local_db.py:1]:
```suggestion
from datetime import datetime
from typing import List, TYPE_CHECKING

from src.connection.dal import DataAccessLayer
from src.models.pre_approval_output import PreApprovalOutput
from tests.factories.pre_approval import PreApprovalFactory
```

File: requirements.txt [hunk 14]
```diff
@@ -0,0 +1,4 @@
+snowflake-connector-python==3.0.2
+snowflake-sqlalchemy==1.4.7
+SQLAlchemy==1.4.47
```
[the remark this answers is not in the document]
Comment (Jvillegasd) [reply requirements.txt:3]: Because snowflake connector is not compatible with SQLAlchemy 2 D:

File: src/models/pre_approval_output.py (hunk 9 above)
Comment (Jvillegasd) [reply src/models/pre_approval_output.py:16]: Because the snowflake table has their fields in UPPER_CASE, so in order to map the table I have to do it

File: tests/factories/pre_approval.py (hunk 10 above)
Comment (Jvillegasd) [reply tests/factories/pre_approval.py:8]: The main reason of why I didn't use UPPER_CASE in this case is because this constants has a pre-calculated value

File: tests/factories/pre_approval.py (hunk 11 above)
Comment (Jvillegasd) [reply tests/factories/pre_approval.py:9]: The main reason of why I didn't use UPPER_CASE in this case is because this constant has a pre-calculated value

File: src/handler.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+def main(event: dict, context: dict) -> dict:
+    response_msg: str = 'Pre approvals sent to Salesforce successfully!'
+    processed_output: WorkflowOutput = process_latest_pre_approvals()
+    processed_batch = processed_output['current_sf_batch']
+
+    if len(processed_batch.loanapp_ids) + len(processed_batch.failed_loanapp_ids) > 0:
+        processed_batch.save()
+        LastSalesforceBatchRegistry.save_batch(
+            dynamodb_table=last_batch_table,
```
[the remark this answers is not in the document]
Comment (Jvillegasd) [reply src/handler.py:72]: I put these variables as global in the starting iterations of the solution, there isn't particular reason for it

File: src/handler.py (hunk 5 above)
Comment (Jvillegasd) [reply src/handler.py:53]: I put this variable as global in the starting iterations of the solution, there isn't particular reason for it

File: serverless.yml [hunk 16]
```diff
@@ -0,0 +1,49 @@
+org: ctsai
```
[the remark this answers is not in the document]
Comment (Jvillegasd) [reply serverless.yml:1]: I think @Mandon24 can help us with this change

File: requirements.txt [hunk 17]
```diff
@@ -0,0 +1,4 @@
...
+typing_extensions==4.5.0
```
[the remark this answers is not in the document]
Comment (Jvillegasd) [reply requirements.txt:4]: It was installed with snowflake connector

**Review by CFCarlo1911 — APPROVED** [review]

File: tests/factories/pre_approval.py (hunk 10 above)
Comment (jjmonsalveg) [reply tests/factories/pre_approval.py:8]: but they are still constant once you calculate them maybe should be something like _NOW

File: tests/factories/pre_approval.py (hunk 11 above)
Comment (jjmonsalveg) [reply tests/factories/pre_approval.py:9]: but they are still constant once you calculate them maybe should be something like _NEXT_ONE_WEEK

File: serverless.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (jjmonsalveg) [reply serverless.yml:1]: I see I think they refer to this serverless dashboard. Because in the specification I am not seeing this as something required. So is this required by and for? 
https://www.serverless.com/framework/docs/providers/aws/guide/serverless.yml

[the remark this answers is not in the document]
Comment (Mandon24) [reply serverless.yml:1]: yes but the org reflects and is tied to the org name on the sls dashboard so i recommend not to change it to caminofinancial unless everything is migrated to a new org name caminofinancial.

[the remark this answers is not in the document]
Comment (jjmonsalveg) [reply serverless.yml:1]: but is it mandatory? I hate(not really) to see the nickname of someone here ctsai 😅

Can we just omit it?

File: src/handler.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+                PreApprovalOutput.CREATED_AT >= last_batch.created_at,
+                PreApprovalOutput.LOANAPP_ID.not_in(last_batch.loanapp_ids)
+            )
+        )
+    for output in outputs_query.filter(*filters).yield_per(batch_size):
+        yield output
+
+
+def process_latest_pre_approvals() -> WorkflowOutput:
+    current_sf_sync_batch: int = 1
```
Comment (alisurti7) [inline src/handler.py:42]: This sf_sync_batch is meant to store the order in which the batches were processed, right? If so, is there a reason why you're not using the created_at field instead?

File: src/handler.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    batch_size: int = 50
+) -> Iterator[PreApprovalOutput]:
+    filters: list = []
+    outputs_query = db.query(PreApprovalOutput)
+
+    if last_batch:
+        filters.append(
+            and_(
+                PreApprovalOutput.CREATED_AT >= last_batch.created_at,
+                PreApprovalOutput.LOANAPP_ID.not_in(last_batch.loanapp_ids)
```
Comment (alisurti7) [inline src/handler.py:34]: Few questions on these filters (please pardon my ignorance if I'm missing something crucial):
Is the created_at value for Preapproval record on snowflake guaranteed to match with the SF batch created at value?
Out of curiosity, why do the loanapp_id check here? Why isn't the created_at filter sufficient? Wouldn't it make more sense to check against loanapp_ids from all batches, not just one?
Should failed loanapp_ids be checked against?

File: serverless.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (Jvillegasd) [reply serverless.yml:1]: As I know, the org name is mandatory, in this way Serverless framework will check on which organization the project is going to be deployed. I think we should check this on later iteration

File: src/handler.py (hunk 18 above)
Comment (Jvillegasd) [reply src/handler.py:42]: Nope, this sync_batch is just for metadata purpose, we are intended to use create_at to check where the batches were processed

File: src/handler.py (hunk 19 above)
Comment (Jvillegasd) [reply src/handler.py:34]: when the algorithm runs, the create_at is calculated and all new Preapproval outputs that exists at the moment are processed, so this will guarantee the matching. The loanapp_id checking is to handle a very tricky edge-case when a new unprocessed Preapproval output has the same create_at date than the one calculated for the SF batch

**Review by CFCarlo1911 — APPROVED** [review]

File: tests/test_handler.py [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+    assert lambda_response['statusCode'] == 200
+
+    batch_id = json.loads(lambda_response['body'])['batch_id']
+    sf_table = dynamodb_resource.Table('SnowflakeBatchRegistryTable')
+    current_sf_batch = sf_table.get_item(Key={'batch_id': batch_id})
+
+    assert 'Item' not in current_sf_batch
+
+
+def test_workflow(
```
Comment (Mandon24) [inline tests/test_handler.py:98]: I suggest taking another look at this test case since this one was flaky for us when we tested.

**Review by Mandon24 — APPROVED** [review]

File: serverless.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (jjmonsalveg) [reply serverless.yml:1]: not cool at all https://www.serverless.com/framework/docs/guides/dashboard , but I guess we can flesh out later how to create a new org or rename the current `ctsai`

File: tests/test_handler.py (hunk 20 above)
Comment (Jvillegasd) [reply tests/test_handler.py:98]: I did some modification in the factory that creates the test objects in order to avoid the unwanted testcase

Comment (jjmonsalveg) [reply tests/test_handler.py:98]: I'm running the current implementation 1k times so far so good but let me check your trick @Jvillegasd

File: conftest.py [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+
+import boto3
+import pytest
+from moto import mock_dynamodb
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
+
+from src.connection.dal import DataAccessLayer, Base
+
+if TYPE_CHECKING:
```
Comment (jjmonsalveg) [inline conftest.py:15]: why do we need this?

Comment (jjmonsalveg) [reply conftest.py:15]: (I test them without the if) I know sometimes this is used to avoid circular dependencies but I still don't get it. Help me to understand it

File: conftest.py [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+
+    dal = DataAccessLayer()
+    dal.snowflake_url = 'sqlite://'
+    dal.engine = create_engine(dal.snowflake_url)
+    dal.session = sessionmaker(
+        autocommit=False,
+        autoflush=False,
+        expire_on_commit=False,
+        bind=dal.engine
+    )
```
Comment (jjmonsalveg) [inline conftest.py:33]: I like the idea to encapsulate the creation of the connection inside of DataAccessLayer but I'll suggest two potential refactors depending on the preference you can choose between dataclass o a dict, but either case I'll suggest passing the config object to `DataAccessLayer`. I'll put my preference here:

```
@dataclass
class ConfigConnection:
    """Class to setup the SqlAlchemy parameters"""

    user: str
    password: str
    account: str
    werehouse: str
    database: str
    schema: str

class DataAccessLayer:
    def __init__(self, config: ConfigConnection):
        self.snowflake_url: str = URL(**asdict(config))
        self.engine = create_engine(self.snowflake_url)
        self.session = sessionmaker(
            autocommit=False, autoflush=False, expire_on_commit=False, bind=self.engine
        )
```

And in the production code, you have one Config object the same for the test

File: serverless.yml [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+provider:
+  name: aws
+  runtime: python3.10
+  environment:
+    SNOWFLAKE_USER: ${env:SNOWFLAKE_USER}
+    SNOWFLAKE_PASSWORD: ${env:SNOWFLAKE_PASSWORD}
+    SNOWFLAKE_ACCOUNT: ${env:SNOWFLAKE_ACCOUNT}
+    SNOWFLAKE_WAREHOUSE: ${env:SNOWFLAKE_WAREHOUSE}
+    SNOWFLAKE_DATABASE: ${env:SNOWFLAKE_DATABASE}
+    SNOWFLAKE_SCHEMA: ${env:SNOWFLAKE_SCHEMA}
```
[the remark this answers is not in the document]
Comment (Jvillegasd) [reply serverless.yml:17]: Got it!

File: src/handler.py [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+    processed_batch = processed_output['current_sf_batch']
+
+    if len(processed_batch.loanapp_ids) + len(processed_batch.failed_loanapp_ids) > 0:
+        processed_batch.save()
+        LastSalesforceBatchRegistry.save_batch(
+            dynamodb_table=last_batch_table,
+            next_sf_batch=processed_batch
+        )
+        if processed_output['last_sf_batch_id']:
+            LastSalesforceBatchRegistry.remove_batch(
```
Comment (jjmonsalveg) [inline src/handler.py:78]: ooh! I see why you delete is because you want to list the last one, but I invite you to think about how to do it without trusting 100% in the delete because suppose the delete fails, does the algorithm break? For now, I think it's ok, maybe have a table with a single element and just update it. Maybe for this use case

File: src/connection/dal.py [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+from contextlib import contextmanager
+from typing import Iterator
+
+from snowflake.sqlalchemy import URL
+from sqlalchemy import create_engine
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.orm.session import Session
+
+Base = declarative_base()
```
Comment (jjmonsalveg) [inline src/connection/dal.py:11]:
```suggestion
BaseModel = declarative_base()
```
(nitpick)

File: src/models/pre_approval_output.py [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+    PRE_APPROVAL_OUTPUT = Column(String)
+    SUMMARY = Column(String)
+    BO_DENIAL_REASON = Column(String)
+    SF_DENIAL_REASON = Column(String)
+
+    # TODO: Support for test
+    if 'pytest' not in sys.modules:
+        MONTHLY_BALANCES = Column(ARRAY)
+
+    # Since snowflake table does not have primary_key, asign a virtual one
```
Comment (jjmonsalveg) [inline src/models/pre_approval_output.py:39]: 🤐 ok definitely there is no way to ensure we won't have duplicates

File: src/schemas/workflow.py [hunk 27]
```diff
@@ -0,0 +1,8 @@
+from typing import TypedDict
+
+from src.models.batch_registry import SalesforceBatchRegistry
+
+
+class WorkflowOutput(TypedDict):
```
Comment (jjmonsalveg) [inline src/schemas/workflow.py:6]:
```suggestion
class ShortHistoryBatch(TypedDict):
```
LatestHistoryBatch, RecentHistoryBatch.... (nitpick)

File: tests/factories/pre_approval.py [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+    PRE_APPROVAL_OUTPUT = factory.Faker('pystr')
+    SUMMARY = factory.Faker('pystr')
+    BO_DENIAL_REASON = factory.Faker('pystr')
+    SF_DENIAL_REASON = factory.Faker('pystr')
+
+    # MONTHLY_BALANCES = factory.LazyAttribute(
+    #     lambda self: [
+    #         factory.Faker('pyfloat', min_value=0, max_value=1000)
+    #     ] * NUMBER_OF_BALANCES
+    # )
```
Comment (jjmonsalveg) [inline tests/factories/pre_approval.py:41]:
```suggestion
```

File: tests/test_local_db.py [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+    )
+
+    assert fetch_response is not None
+    assert 'Item' in fetch_response
+    assert fetch_response['Item']['batch_id'] == 'test_id'
+
+
+def test_sqlite_table_put_item(sqlite_dal: DataAccessLayer):
+    with sqlite_dal.get_session() as db:
+        factory_data: List[PreApprovalOutput] = PreApprovalFactory.build_batch(size=10)
```
Comment (jjmonsalveg) [inline tests/test_local_db.py:54]:
```suggestion
def test_sqlite_table_put_item(sqlite_dal: DataAccessLayer):
    with sqlite_dal.get_session() as db:
        batch_size = 3
        factory_data: List[PreApprovalOutput] = PreApprovalFactory.build_batch(size=batch_size)
```

File: tests/test_handler.py [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+    current_sf_batch = sf_table.get_item(Key={'batch_id': batch_id})
+
+    assert 'Item' in current_sf_batch
+
+    assert int(current_sf_batch['Item']['sync_batch']) == 1
+    assert len(factory_data) == len(current_sf_batch['Item']['loanapp_ids'])
+    assert all(
+        factory_item.LOANAPP_ID == int(loanapp_id)
+        for factory_item, loanapp_id in zip(factory_data, current_sf_batch['Item']['loanapp_ids'])
+    )
```
Comment (jjmonsalveg) [inline tests/test_handler.py:69]: ❤️

**Review by jjmonsalveg — COMMENTED** [review]
AWESOME PR, really good code. Let's ask the question to Data team about duplicates, I think your code solve the difficult problem if they decide to relax the delta algorithm should be easier and maybe we can do an simplification.

File: conftest.py (hunk 21 above)
Comment (Jvillegasd) [reply conftest.py:19]: This conditional is because we are using boto3-suds for boto3 type hinting. So, in order to reduce lambda size, we excluded this lib form deployment package. If we removed the conditional, the lambda will explode in prod

**Review by jjmonsalveg — APPROVED** [review]

**Review by Mandon24 — APPROVED** [review]
